support Ford Bronco Sport 2021-22, Maverick 2022

incognitojam opened the pull request:
- [x] new models
  - Bronco Sport 2021+
  - Maverick 2022+
- [x] test resume from stop works properly
- [x] check that TJA is toggled off
- [x] radar parser disabled because it's buggy
- [x] make steering good (ish)
  - https://github.com/commaai/cereal/pull/381
  
Note: In `LatCtlCurv_No_Actl`, positive curvature means left

incognitojam committed:
cleanup

incognitojam committed:
cleanup steering command

incognitojam committed:
disable radar

incognitojam committed:
FordCAN refactor

incognitojam committed:
add Ford Bronco Sport 2021 fw

VIN: 3FMCR9B61MRA17853

62241b0c7fea4589|2022-10-25--13-40-05--0

incognitojam committed:
curvature steering

sshane commented:
Not sure if you were already going to, but feel free to split some of the refactors into smaller mergable PRs so you can reduce this diff

incognitojam committed:
keep command angle close to current steering angle

incognitojam committed:
refactor angle rate limits

adeebshihadeh committed:
will it steer?

incognitojam committed:
cleanup latcontrol_angle

incognitojam committed:
add curvature lateral controller

incognitojam committed:
bump cereal (curvature)

incognitojam committed:
implement Ford controller

incognitojam commented on selfdrive/controls/lib/latcontrol_curvature.py:
```diff
@@ -0,0 +1,43 @@
+      curvature_log.i = self.pid.i
+      curvature_log.d = self.pid.d
+      curvature_log.f = self.pid.f
+      curvature_log.output = -output
```
need to calculate saturated?

incognitojam commented on selfdrive/car/ford/carcontroller.py:
```diff
@@ -46,33 +30,35 @@ def update(self, CC, CS):
+        lca_rq = 1
+
+        # use LatCtlCurv_No_Actl to actuate steering
+        # TODO: apply rate limits
```
add rate limits back

incognitojam committed:
fix range

incognitojam commented:
Route on `ford-devel` which is based on this branch: `26b2cace68e36212|2022-11-05--20-51-49`

incognitojam committed:
fordcan: rename tja msg to lat_ctl

incognitojam committed:
fordcan: rename lca_rq to mode

incognitojam committed:
ford: add Escape learned steer ratio

incognitojam committed:
Ford: add Maverick 2022

VIN: 3FTTW8E36NRA74219

a6427650bff6b419|2022-09-19--10-51-30

incognitojam committed:
fix ramp type (deadzone)

incognitojam committed:
latcontrol curvature: add angle saturation check

incognitojam commented on selfdrive/controls/lib/latcontrol_curvature.py:
```diff
@@ -0,0 +1,50 @@
+class LatControlCurvature(LatControl):
+  def __init__(self, CP, CI):
+    super().__init__(CP, CI)
+    self.steer_max = 0.02
```
this feels like a hack. I thought to use [-1.0, 1.0] and apply in carcontroller, but I think would have to test new PID coefficients?

incognitojam committed:
clip curvature

incognitojam committed:
remove

incognitojam commented:
another route on `ford-devel` (rebased on this branch):
`26b2cace68e36212|2022-11-21--23-21-12`

sshane commented on selfdrive/controls/lib/latcontrol_curvature.py:
```diff
@@ -0,0 +1,49 @@
+from selfdrive.controls.lib.latcontrol import LatControl, MIN_STEER_SPEED
+from selfdrive.controls.lib.pid import PIDController
+
+STEER_ANGLE_SATURATION_THRESHOLD = 2.5  # Degrees
```
move to latcontrol.py and import

sshane commented on selfdrive/controls/lib/latcontrol_curvature.py:
```diff
@@ -0,0 +1,49 @@
+    super().__init__(CP, CI)
+    self.pid = PIDController((CP.lateralTuning.pid.kpBP, CP.lateralTuning.pid.kpV),
+                             (CP.lateralTuning.pid.kiBP, CP.lateralTuning.pid.kiV),
+                             k_f=CP.lateralTuning.pid.kf, pos_limit=self.steer_max, neg_limit=-self.steer_max)
```
```suggestion
                             k_f=CP.lateralTuning.pid.kf)
```

Let's leave it up to the PID defaults (practically unlimited). Any number seems arbitrary to me. 0.02 is 20 m/s^2 of lateral accel at 70 mph, and 1.0 (steer_max) is similar in that vein.

sshane commented on selfdrive/controls/controlsd.py:
```diff
@@ -608,9 +611,10 @@ def state_control(self, CS):
-      actuators.steer, actuators.steeringAngleDeg, lac_log = self.LaC.update(CC.latActive, CS, self.VM, lp,
-                                                                             self.last_actuators, self.steer_limited, self.desired_curvature,
-                                                                             self.desired_curvature_rate, self.sm['liveLocationKalman'])
+      actuators.steer, actuators.steeringAngleDeg, actuators.curvature, lac_log = \
+        self.LaC.update(CC.latActive, CS, self.VM, lp,
+                        self.last_actuators, self.steer_limited, self.desired_curvature,
+                        self.desired_curvature_rate, self.sm['liveLocationKalman'])
```
```suggestion
      actuators.steer, actuators.steeringAngleDeg, actuators.curvature, lac_log = \
        self.LaC.update(CC.latActive, CS, self.VM, lp,
                        self.last_actuators, self.steer_limited, self.desired_curvature,
                        self.desired_curvature_rate, self.sm['liveLocationKalman'])
```

We might want to start passing in `actuators` and the lateral controller can modify it

incognitojam committed:
remove PID limits

Co-authored-by: Shane Smiskol <shane@smiskol.com>

incognitojam committed:
move steer angle staturation threshold constant to LatControl

incognitojam committed:
controlsd: pass actuators to LatControl to modify

incognitojam commented on selfdrive/controls/controlsd.py:
```diff
@@ -620,7 +627,7 @@ def state_control(self, CS):
           steer = clip(self.sm['testJoystick'].axes[1], -1, 1)
           # max angle is 45 for angle-based cars
-          actuators.steer, actuators.steeringAngleDeg = steer, steer * 45.
+          actuators.steer, actuators.steeringAngleDeg, actuators.curvature = steer, steer * 45., steer * 0.02
```
maybe we could do something smarter here

incognitojam commented on selfdrive/controls/lib/latcontrol_pid.py:
```diff
@@ -45,4 +45,5 @@ def update(self, active, CS, VM, params, last_actuators, steer_limited, desired_
       pid_log.output = output_steer
       pid_log.saturated = self._check_saturation(self.steer_max - abs(output_steer) < 1e-3, CS, steer_limited)
 
-    return output_steer, angle_steers_des, pid_log
```
I didn't set `actuators.steeringAngleDeg` because we already set `pid_log.steeringAngleDesiredDeg`

incognitojam replied:
oh this might be why process replay changed

incognitojam commented on selfdrive/controls/controlsd.py:
```diff
@@ -642,6 +649,8 @@ def state_control(self, CS):
         if self.CP.steerControlType == car.CarParams.SteerControlType.angle:
           steering_value = actuators.steeringAngleDeg
+        elif self.CP.steerControlType == car.CarParams.SteerControlType.curvature:
+          steering_value = actuators.curvature
```
User reported that "steering limit reached" alert never appears - this line is suspect but I tried it both `-ve` and `+ve` and it didn't appear. maybe the issue is somewhere else.

incognitojam commented on selfdrive/car/tests/routes.py:
```diff
@@ -16,8 +16,10 @@
 
 # TODO: add routes for these cars
 non_tested_cars = [
+  FORD.BRONCO_SPORT_MK1,
```
@adeebshihadeh We have some routes for the Bronco Sport, maybe we could use one of those?

incognitojam committed:
cleanup imports

incognitojam committed:
fix curvature control signs

incognitojam committed:
Ford: positive curvature means left!

incognitojam committed:
remove comments

incognitojam committed:
Ford: add missing Maverick 2022 FW

VIN: 3FTTW8F98NRA75889

c845bd8c366e6f3d|2022-12-03--18-57-00--0

incognitojam committed:
fix regen

sshane commented on selfdrive/controls/lib/latcontrol_curvature.py:
```diff
@@ -0,0 +1,48 @@
+from selfdrive.controls.lib.pid import PIDController
+
+
+class LatControlCurvature(LatControl):
```
you can subclass `LatControlPID` and remove the init function here

sshane replied:
you can subclass again if you do something like https://github.com/commaai/openpilot/pull/26744/commits/d8ada72314ed2990527e3c13be1ae4cb1528b08f

incognitojam committed:
inherit LatControlPID

adeebshihadeh commented on selfdrive/car/ford/interface.py:
```diff
@@ -1,40 +1,52 @@
-    ret.steerControlType = CarParams.SteerControlType.angle
-    ret.steerActuatorDelay = 0.4
-    ret.steerLimitTimer = 1.0
+    # These cars are dashcam only until steering safety is implemented
```
```suggestion
    # These cars are dashcam only until the port is finished
```

adeebshihadeh committed:
Update selfdrive/car/ford/interface.py

adeebshihadeh commented on selfdrive/car/ford/fordcan.py:
```diff
@@ -4,161 +4,164 @@
-      lines = 3  # WarnLeft_NoRight
-    elif hud_control.rightLaneDepart:
-      lines = 15  # NoLeft_WarnRight
+class FordCAN:
```
any reason this is a class?

incognitojam replied:
just to remove packer arg, and then when we get around to CAN-FD we will need to specify a different bus too

adeebshihadeh replied:
i'd rather not introduce new patterns here

incognitojam replied:
oh I just saw that tesla already did it https://github.com/commaai/openpilot/blob/master/selfdrive/car/tesla/teslacan.py

incognitojam replied:
I can revert for now

adeebshihadeh commented on selfdrive/car/ford/carcontroller.py:
```diff
@@ -46,33 +31,33 @@ def update(self, CC, CS):
-      path_angle = math.radians(apply_angle) / CarControllerParams.LKAS_STEER_RATIO
+      if CC.latActive:
+        # use LatCtlCurv_No_Actl to actuate steering
+        # IMPORTANT! left is positive
```
keep comments like this in one place. right now, some are here and others in fordcan.py

incognitojam committed:
move signal positive/negative comments to FordCAN

incognitojam committed:
revert FordCAN and cleanup

incognitojam committed:
remove d gain

sshane commented on selfdrive/controls/lib/latcontrol_curvature.py:
```diff
@@ -1,21 +1,11 @@
-    super().__init__(CP, CI)
-    self.pid = PIDController((CP.lateralTuning.pid.kpBP, CP.lateralTuning.pid.kpV),
-                             (CP.lateralTuning.pid.kiBP, CP.lateralTuning.pid.kiV),
-                             k_f=CP.lateralTuning.pid.kf)
```
Oh I forgot now the PID limits (eg. `pos_limit`) are set in LatControlPID

incognitojam committed:
Merge branch 'master' into ford-now

incognitojam committed:
Revert "inherit LatControlPID"

This reverts commit 62cd235713cd62e98b7759c1fc69356ccf22004a.

incognitojam committed:
bump cereal

incognitojam committed:
cleanup

incognitojam committed:
Merge remote-tracking branch 'origin/master' into ford-now

incognitojam committed:
inherit LatControlPID

incognitojam committed:
Merge remote-tracking branch 'origin/master' into ford-now

incognitojam committed:
rename LatCtl mode to lca_rq

incognitojam committed:
LatControlCurvature: fix desired steering angle/saturation check

incognitojam committed:
fix actuatorsOutput curvature

incognitojam committed:
add Lincoln Aviator 2021 to Ford Explorer platform

Based on the Ford Explorer, the Aviator has very similar firmware
versions. Add these to the Explorer platform and create a new CarInfo
for the Aviator.

f0709d2bc6ca451f|2022-12-10--12-36-59--0

VIN: 5LM5J7XC8MGL09541

incognitojam committed:
Merge branch 'master' into ford-now

incognitojam committed:
fix desired steering angle calculation (for ramp rate)

incognitojam committed:
increase steerActuatorDelay

incognitojam committed:
update Ford Escape steer ratio

incognitojam committed:
update Ford Focus steer ratio

incognitojam committed:
update Ford Maverick steer ratio

incognitojam committed:
Ford: add curvature rate limits

incognitojam committed:
Ford: new gear shifter signal to support Aviator

incognitojam committed:
Merge branch 'master' into ford-now

incognitojam committed:
Ford: add Maverick 2022 XLT fw

VIN: `3FTTW8E39NRB00148`

933ad9d04ac82693|2023-01-04--17-31-33--0

incognitojam committed:
formatting

incognitojam committed:
bump steerLimitTimer to default 1.0s

incognitojam committed:
Merge branch 'master' into ford-now

incognitojam committed:
tweaks

incognitojam committed:
Merge branch 'master' into ford-now

incognitojam committed:
no dashcam only

incognitojam committed:
add Ford Bronco Sport 2022 fw

VIN: 3FMCR9B69NRD15990

54827bf84c38b14f|2023-01-23--14-58-23--0

incognitojam committed:
bronco sport 2022 car info

incognitojam committed:
generate docs

incognitojam committed:
Revert "generate docs"

This reverts commit 85195d2167310ac36f3dee075016590dec42ce92.

incognitojam committed:
Revert "no dashcam only"

This reverts commit 5f57cd11ef13f9b8d119da484ba8aa9754e1e2ad.

incognitojam committed:
Merge branch 'master' into ford-now

sshane committed:
simplify things

sshane committed:
fix

sshane committed:
revert non-ford stuff

sshane committed:
revert cc newline changes

sshane commented on selfdrive/car/ford/carstate.py:
```diff
@@ -14,7 +14,7 @@ def __init__(self, CP):
     can_define = CANDefine(DBC[CP.carFingerprint]["pt"])
     if CP.transmissionType == TransmissionType.automatic:
-      self.shifter_values = can_define.dv["Gear_Shift_by_Wire_FD1"]["TrnGear_D_RqDrv"]
+      self.shifter_values = can_define.dv["Gear_Shift_by_Wire_FD1"]["TrnRng_D_RqGsm"]
```
Can we do all this extra stuff in another PR? Trying to make this PR as small as possible

incognitojam committed:
Revert "Ford: new gear shifter signal to support Aviator"

This reverts commit 673a05e511fe2ee7545c116cc630a2bfffd985e1.

incognitojam committed:
Revert "add Lincoln Aviator 2021 to Ford Explorer platform"

This reverts commit ac6e4e86af5714c160f909c73c5ff33328986e64.

sshane commented:
Split into these PRs:
- https://github.com/commaai/openpilot/pull/27090
- https://github.com/commaai/openpilot/pull/27092
- https://github.com/commaai/openpilot/pull/27094
- https://github.com/commaai/openpilot/pull/27093
- https://github.com/commaai/openpilot/pull/27091
- https://github.com/commaai/openpilot/pull/27096
- https://github.com/commaai/openpilot/pull/27095
